Replace angular `toastr` with preact-compatible toast messages

lyzadanger opened the pull request:
This PR implements toast messages via a preact-compatible, more accessible mechanism.

`ToastMessage` components are rendered as list elements inside of a `ul` (corresponding to a `ToastMessages` component) that is an `aria-live` region. The `toastMessenger` service handles the timing of how long a message is present and handles "dismissing" it. The store module involved here is relatively "dumb", by design, and just maintains and returns messages in state.

This PR has two commits, **which could be split into two PRs if that would aid review**:

* https://github.com/hypothesis/client/commit/a4f650ac270edadccab3a5dd1578ee22f0b95080 : implementation: `toastMessenger` service, `toastMessages` store module, `toastMessages` component—not hooked up to anything in this commit.
* https://github.com/hypothesis/client/commit/a75df7125b45d1384586483a5553253f17f74956 : update preact components to use `toastMessenger` service

The old `flash` service needs to be pulled out of a few remaining places and removed from the codebase, but this diff is getting quite large, so I will leave those remaining spots for a follow-up.

The implementation is feeling mostly solid; I want to get to a shippable point and then we can tweak animations, colors, etc., as we see fit.

At present, messages look like this:

Error:

![image](https://user-images.githubusercontent.com/439947/77194931-36d96e00-6ab7-11ea-82ba-f97b5e4cc9b0.png)

Success:

![image](https://user-images.githubusercontent.com/439947/77194950-3e007c00-6ab7-11ea-8957-703ba3f09e7b.png)


Updates since last WIP PR:

* There is now an interim, "dismissed" state before a message is removed from the store/DOM
* Tests are in place

Fixes https://github.com/hypothesis/client/issues/1925

LMS007 commented on src/sidebar/store/modules/toast-messages.js:
```diff
@@ -0,0 +1,92 @@
+      if (message.id && message.id === action.message.id) {
+        return { ...action.message };
+      }
+      return { ...message };
```
I'm okay with this, but my preference is not to make a copy of the thing if its not changing only so its clear in code when we do make a copy, we are in fact making changes.

lyzadanger replied:
You're right, I agree.

LMS007 commented on src/sidebar/store/modules/toast-messages.js:
```diff
@@ -0,0 +1,92 @@
+  return { type: actions.UPDATE_MESSAGE, message };
+}
+
+/** Selectors */
```
I like this comment, can you add one for "Actions" above?

lyzadanger replied:
I'm going to do this—I'm also going to add a discussion topic for our next frontend sync about the appropriate term for these functions, as they aren't truly actions, right?

LMS007 commented on src/sidebar/components/test/toast-messages-test.js:
```diff
@@ -0,0 +1,146 @@
+import mockImportedComponents from '../../../test-util/mock-imported-components';
+
+import ToastMessages, { $imports } from '../toast-messages';
+//import { checkAccessibility } from '../../../test-util/accessibility';
```
can be removed

lyzadanger replied:
Actually, I'd argue instead that I forgot to add `a11y` tests, which I'll do now! I wonder if there's something automated that could catch when we forget to add these tests.

lyzadanger replied:
...and they kicked up something legit, so, yay!

LMS007 commented on src/sidebar/components/test/toast-messages-test.js:
```diff
@@ -0,0 +1,146 @@
+    [
+      { message: fakeSuccessMessage(), icon: 'success' },
+      { message: fakeErrorMessage(), icon: 'error' },
+    ].forEach(testCase => {
```
can we have a third test with more than 1 message?

LMS007 commented on src/sidebar/services/test/toast-messenger-test.js:
```diff
@@ -0,0 +1,157 @@
+      service.success('hooray');
+
+      // Move to the first scheduled timeout, which should invoke the
+      // `dismiss` method
```
👍 thumbs

LMS007 commented on src/sidebar/services/test/toast-messenger-test.js:
```diff
@@ -0,0 +1,157 @@
+import toastMessenger from '../toast-messenger';
+
+describe('toastMessenger', function() {
```
Honestly, I forgot our standard for this, and perhaps we need to resist in the next front end sync. For services with multiple exports, should we have this the full file path or just the server file name or the service name?
"sidebar/services/toastMessenger" for example? I think in another PR, we may do a clean up of all test files in services/  I'm happy to take that on.

lyzadanger commented on src/sidebar/components/toast-messages.js:
```diff
@@ -0,0 +1,93 @@
+import SvgIcon from './svg-icon';
+
+/**
+ * An individual toast message—a brief and transient success or error message.
```
— is a long dash, just a bit of punctuation. I could change it to a colon (`:`) if that would be clearer. It separates the clause that defines what a toast message is :).

lyzadanger commented on src/sidebar/components/toast-messages.js:
```diff
@@ -0,0 +1,93 @@
+  return (
+    /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions */
+    <li
+      className={classnames('toast-message-container', {
```
I don't think those roles are accurate for this element, though—it contains something that isn't just for presentation or none...

lyzadanger commented on src/sidebar/services/toast-messenger.js:
```diff
@@ -0,0 +1,71 @@
+      return;
+    }
+
+    const id = generateHexString(10);
```
I want the `id` to be unique because it's used as a (preact index) key on the rendered `ToastMessage` component, but your idea is otherwise smart!

lyzadanger commented on src/sidebar/services/toast-messenger.js:
```diff
@@ -0,0 +1,71 @@
+import { generateHexString } from '../util/random';
+
+const MESSAGE_DISPLAY_TIME = 5000;
+const MESSAGE_DISMISS_DELAY = 1000;
```
How about we split the difference at 500ms? :)

lyzadanger commented on src/sidebar/services/toast-messenger.js:
```diff
@@ -0,0 +1,71 @@
+   * it after a bit. This allows effects/animations to happen before the
+   * message is removed entirely.
+   */
+  const dismiss = messageId => {
```
Thanks! Good feedback. I'll update the methods that have any sort of ambiguity.

lyzadanger commented on src/sidebar/store/modules/toast-messages.js:
```diff
@@ -0,0 +1,92 @@
+const update = {
+  ADD_MESSAGE: function(state, action) {
+    return {
+      messages: state.messages.concat([{ ...action.message }]),
```
Oh, derp.

lyzadanger commented:
I've made some updates to tests:

* Added accessibility test to `ToastMessages` component
* Had a couple of untested things in the store module. Fixed.
